MoreLinq/GroupPartition.cs

```diff
@@ -0,0 +1,144 @@
+            TKey k1, TKey k2, IEqualityComparer<TKey> comparer,
+            Func<IEnumerable<TElement>, IEnumerable<TElement>, IEnumerable<IGrouping<TKey, TElement>>, TResult> resultSelector) =>
+            PartitionImpl(source, 2, k1, k2, default(TKey), comparer, (a, b, c, rest) => resultSelector(a, b, rest));
+
+        /// <summary>
+        /// Partitions a grouping into a projection of elements matching three
+        /// sets of keys and those groups that do not.
+        /// </summary>
+
+        public static TResult Partition<TKey, TElement, TResult>(this IEnumerable<IGrouping<TKey, TElement>> source,
+            TKey k1, TKey k2, TKey k3,
+            Func<IEnumerable<TElement>, IEnumerable<TElement>, IEnumerable<TElement>, IEnumerable<IGrouping<TKey, TElement>>, TResult> resultSelector) =>
+            Partition(source, k1, k2, k3, null, resultSelector);
+
+        /// <summary>
+        /// Partitions a grouping into a projection of elements matching three
+        /// sets of keys and those groups that do not. An additional parameter
+        /// specifies how to compare keys for equality.
+        /// </summary>
+
+        public static TResult Partition<TKey, TElement, TResult>(this IEnumerable<IGrouping<TKey, TElement>> source,
+            TKey k1, TKey k2, TKey k3, IEqualityComparer<TKey> comparer,
+            Func<IEnumerable<TElement>, IEnumerable<TElement>, IEnumerable<TElement>, IEnumerable<IGrouping<TKey, TElement>>, TResult> resultSelector) =>
+            PartitionImpl(source, 3, k1, k2, k3, comparer, resultSelector);
+
+        static TResult PartitionImpl<TKey, TElement, TResult>(IEnumerable<IGrouping<TKey, TElement>> source,
+            int count, TKey k1, TKey k2, TKey k3, IEqualityComparer<TKey> comparer,
+            Func<IEnumerable<TElement>, IEnumerable<TElement>, IEnumerable<TElement>, IEnumerable<IGrouping<TKey, TElement>>, TResult> resultSelector)
+        {
+            Debug.Assert(count > 0 && count <= 3);
+
+            if (source == null) throw new ArgumentNullException(nameof(source));
+            if (resultSelector == null) throw new ArgumentNullException(nameof(resultSelector));
+
+            comparer = comparer ?? EqualityComparer<TKey>.Default;
+
+            List<IGrouping<TKey, TElement>> etc = null;
+
+            var groups = new[]
+            {
+                Enumerable.Empty<TElement>(),
+                Enumerable.Empty<TElement>(),
+                Enumerable.Empty<TElement>(),
+            };
+
+            foreach (var e in source)
+            {
+                var i = count > 0 && comparer.Equals(e.Key, k1) ? 0
+                      : count > 1 && comparer.Equals(e.Key, k2) ? 1
+                      : count > 2 && comparer.Equals(e.Key, k3) ? 2
+                      : -1;
+
+                if (i < 0)
+                    (etc ?? (etc = new List<IGrouping<TKey, TElement>>())).Add(e);
+                else
+                    groups[i] = e;
```

Comment:
Good point but isn't it better to turn it into an error? Although doing so for non-matching groups would be unnecessarily expensive. I'm inclined to document this as undefined behaviour even if someone observes what you say. I guess the best test is what you'd expect `Partition` to do if the came from `GroupAdjacent` instead of `GroupBy`. Have to think about this one.